client/src/wayland_obj/layer_shell.rs

```diff
@@ -0,0 +1,168 @@
+    zwlr_layer_shell_v1::{Layer, ZwlrLayerShellV1 as WlrLayerShell},
+    zwlr_layer_surface_v1::{self as wlr_layer_surface, ZwlrLayerSurfaceV1 as WlrLayerSurface}
+};
+
+use crate::area::{Origin, Size};
+
+use super::{Buffer, Output, Surface};
+
+/// The minimum version of the wl_layer_shell protocol to bind to.
+pub(super) const WLR_LAYER_SHELL_VERSION: u32 = 1;
+
+/// LayerShell
+pub(super) struct LayerShellManager {
+    shell: Option<WlrLayerShell>
+}
+
+impl LayerShellManager {
+    pub fn new() -> Self {
+        LayerShellManager { shell: None }
+    }
+
+    pub fn create_layer_surface(
+        &self,
+        wl_surface: Surface,
+        output: Option<&Output>,
+        layer: Layer
+    ) -> Result<LayerSurface, ()> {
+        use wayland_client::Interface; // for ::NAME
+        trace!("create_layer_surface");
+        let shell = self.shell.as_ref().expect(&format!(
+            "No LayerShell registered. Make sure your compositor supports the {} protocol",
+            WlrLayerSurface::NAME
+        ));
+
+        let layer_surface = shell.get_layer_surface(
+            &wl_surface,
+            output.map(AsRef::as_ref),
+            layer,
+            Default::default(),
+            |new_layer_surface| {
+                let state = Rc::new(RefCell::new(LayerSurfaceState::new(wl_surface.clone())));
+                new_layer_surface.implement(LayerSurfaceEventHandler {}, state)
+            }
+        )?;
+
+        let res = LayerSurface::new(layer_surface);
+
+        trace!("committing layersurface");
+        res.commit();
+
+        Ok(res)
+    }
+}
+
+impl GlobalImplementor<WlrLayerShell> for LayerShellManager {
+    fn new_global(&mut self, new_shell: NewProxy<WlrLayerShell>) -> WlrLayerShell {
+        let res = new_shell.implement_dummy();
+        if self.shell.replace(res.clone()).is_some() {
+            use wayland_client::Interface;
+            panic!(
+                "{} already registered. Expected unique global",
+                WlrLayerShell::NAME
+            );
+        }
+        res
+    }
+}
+
+/// LayerSurface
+pub struct LayerSurface {
+    layer_surface: WlrLayerSurface
+}
+
+// Since most properties of WlrLayerSurface are double-buffered, only forward requests
+impl LayerSurface {
+    pub fn new(layer_surface: WlrLayerSurface) -> Self {
+        Self { layer_surface }
+    }
+
+    pub fn set_buffer(&mut self, buffer: &Buffer, origin: Origin) {
+        let mut state = unwrap_state(self.as_ref()).borrow_mut();
+        if let Some((_old_buffer, _)) = state.buffer.replace((buffer.buffer.clone(), origin)) {
+            // TODO(ried): cleanup?
+            warn!("Replacing uncommitted buffer, cleanup?");
+        }
+        // Assigning the Buffer before the first configure event is a protocol error.
+        if state.configuration_serial.is_some() {
+            state.attach_buffer();
+        } else {
+            info!("LayerSurface is not yet configured. Postponing layer attachment");
```

Comment:
This should probably be bumped up to an `error!`, as otherwise this could be lost in the noise.

Also this should probably be represented as an error on the Lua side too, so it should eventually bubble up to an `rlua::Error` (potentially a different error could be returned here, but ultimately to communicate to Lua it must be an `rlua::Error`)